File: src/vodoo/_domain.py
"""Base class for domain-specific namespaces on OdooClient.

Each Odoo model domain (helpdesk, CRM, project tasks, etc.) is exposed as a
namespace on the client::

    client.helpdesk.list(limit=10)
    client.helpdesk.get(42)
    client.helpdesk.comment(42, "Deployed to staging")

Subclasses set class-level attributes (``_model``, ``_default_fields``, etc.)
and optionally override or add domain-specific methods (e.g. ``create``).
"""

from __future__ import annotations

import base64
import builtins
from pathlib import Path
from typing import TYPE_CHECKING, Any, ClassVar

from vodoo.auth import message_post_sudo
from vodoo.exceptions import RecordNotFoundError

if TYPE_CHECKING:
    from vodoo.client import OdooClient


class DomainNamespace:
    """Sync domain namespace — common CRUD, messaging, tags, and attachments.

    Subclasses must set at least ``_model`` and ``_default_fields``.
    """

    _model: ClassVar[str]
    _tag_model: ClassVar[str | None] = None
    _default_fields: ClassVar[list[str]]
    _default_detail_fields: ClassVar[list[str] | None] = None
    _record_type: ClassVar[str] = "Record"

    # Shared field lists (match base.py constants)
    _TAG_FIELDS: ClassVar[list[str]] = ["id", "name", "color"]
    _MESSAGE_FIELDS: ClassVar[list[str]] = [
        "id",
        "date",
        "author_id",
        "body",
        "subject",
        "message_type",
        "subtype_id",
        "email_from",
    ]
    _ATTACHMENT_LIST_FIELDS: ClassVar[list[str]] = [
        "id",
        "name",
        "file_size",
        "mimetype",
        "create_date",
    ]
    _ATTACHMENT_READ_FIELDS: ClassVar[list[str]] = ["name", "datas"]

    def __init__(self, client: OdooClient) -> None:
        self._client = client

    # -- CRUD ----------------------------------------------------------------

    def list(
        self,
        domain: builtins.list[Any] | None = None,
        limit: int | None = 50,
        fields: builtins.list[str] | None = None,
        order: str = "create_date desc",
    ) -> builtins.list[dict[str, Any]]:
        """Search and return records.

        Args:
            domain: Odoo domain filter.
            limit: Max records to return.
            fields: Fields to fetch (``None`` → default fields).
            order: Sort order expression.

        Returns:
            List of record dictionaries.
        """
        if fields is None:
            fields = list(self._default_fields)
        return self._client.search_read(
            self._model,
            domain=domain,
            fields=fields,
            limit=limit,
            order=order,
        )

    def get(
        self,
        record_id: int,
        fields: builtins.list[str] | None = None,
    ) -> dict[str, Any]:
        """Read a single record by ID.

        Args:
            record_id: Record ID.
            fields: Fields to read (``None`` → all fields).

        Returns:
            Record dictionary.

        Raises:
            RecordNotFoundError: If no record with that ID exists.
        """
        if fields is None and self._default_detail_fields is not None:
            fields = list(self._default_detail_fields)
        records = self._client.read(self._model, [record_id], fields=fields)
        if not records:
            raise RecordNotFoundError(self._model, record_id)
        return records[0]

    def set(
        self,
        record_id: int,
        values: dict[str, Any],
    ) -> bool:
        """Update fields on a record.

        Args:
            record_id: Record ID.
            values: Field names → new values.

        Returns:
            ``True`` on success.
        """
        return self._client.write(self._model, [record_id], values)

    def fields(self) -> dict[str, Any]:
        """Return all field definitions for this model."""
        result: dict[str, Any] = self._client.execute(self._model, "fields_get")
        return result

    # -- Messaging -----------------------------------------------------------

    def comment(
        self,
        record_id: int,
        message: str,
        user_id: int | None = None,
        markdown: bool = True,
    ) -> bool:
        """Post a customer-visible comment on a record.

        Args:
            record_id: Record ID.
            message: Comment text (plain or markdown).
            user_id: Post as this user (``None`` → configured default).
            markdown: Convert markdown to HTML before posting.

        Returns:
            ``True`` on success.
        """
        body = _convert_to_html(message, markdown)
        return message_post_sudo(
            self._client,
            self._model,
            record_id,
            body,
            user_id=user_id,
            is_note=False,
        )

    def note(
        self,
        record_id: int,
        message: str,
        user_id: int | None = None,
        markdown: bool = True,
    ) -> bool:
        """Post an internal note (not visible to customers).

        Args:
            record_id: Record ID.
            message: Note text (plain or markdown).
            user_id: Post as this user (``None`` → configured default).
            markdown: Convert markdown to HTML before posting.

        Returns:
            ``True`` on success.
        """
        body = _convert_to_html(message, markdown)
        return message_post_sudo(
            self._client,
            self._model,
            record_id,
            body,
            user_id=user_id,
            is_note=True,
        )

    def messages(
        self,
        record_id: int,
        limit: int | None = None,
    ) -> builtins.list[dict[str, Any]]:
        """List chatter messages for a record.

        Args:
            record_id: Record ID.
            limit: Max messages (``None`` → all).

        Returns:
            List of message dictionaries.
        """
        domain: builtins.list[Any] = [
            ("model", "=", self._model),
            ("res_id", "=", record_id),
        ]
        return self._client.search_read(
            "mail.message",
            domain=domain,
            fields=self._MESSAGE_FIELDS,
            order="date desc",
            limit=limit,
        )

    # -- Tags ----------------------------------------------------------------

    def tags(self) -> builtins.list[dict[str, Any]]:
        """List available tags for this domain.

        Returns:
            List of tag dictionaries.

        Raises:
            ValueError: If the domain has no tag model.
        """
        if self._tag_model is None:
            msg = f"No tag model defined for {self._model}"
            raise ValueError(msg)
        return self._client.search_read(
            self._tag_model,
            fields=self._TAG_FIELDS,
            order="name",
        )

    def add_tag(
        self,
        record_id: int,
        tag_id: int,
    ) -> bool:
        """Add a tag to a record (idempotent).

        Args:
            record_id: Record ID.
            tag_id: Tag ID.

        Returns:
            ``True`` on success.
        """
        record = self.get(record_id, fields=["tag_ids"])
        current_tags: builtins.list[int] = record.get("tag_ids", [])
        if tag_id not in current_tags:
            current_tags.append(tag_id)
            return self._client.write(
                self._model,
                [record_id],
                {"tag_ids": [(6, 0, current_tags)]},
            )
        return True

    # -- Attachments ---------------------------------------------------------

    def attachments(
        self,
        record_id: int,
    ) -> builtins.list[dict[str, Any]]:
        """List attachments on a record.

        Args:
            record_id: Record ID.

        Returns:
            List of attachment metadata dictionaries.
        """
        domain: builtins.list[Any] = [
            ("res_model", "=", self._model),
            ("res_id", "=", record_id),
        ]
        return self._client.search_read(
            "ir.attachment",
            domain=domain,
            fields=self._ATTACHMENT_LIST_FIELDS,
        )

    def attach(
        self,
        record_id: int,
        file_path: Path | str | None = None,
        *,
        data: bytes | None = None,
        name: str | None = None,
    ) -> int:
        """Attach a file (from disk or bytes) to a record.

        Args:
            record_id: Record ID.
            file_path: Path to file on disk (mutually exclusive with *data*).
            data: Raw bytes (mutually exclusive with *file_path*).
            name: Attachment name (defaults to filename; required with *data*).

        Returns:
            ID of the created ``ir.attachment``.

        Raises:
            ValueError: If arguments are invalid.
            FileNotFoundError: If *file_path* does not exist.
        """
        from vodoo.base import _prepare_attachment_upload

        values = _prepare_attachment_upload(file_path, data, name, self._model, record_id)
        return self._client.create("ir.attachment", values)

    def download(
        self,
        record_id: int,
        output_dir: Path | None = None,
        extension: str | None = None,
    ) -> builtins.list[Path]:
        """Download all attachments for a record to disk.

        Args:
            record_id: Record ID.
            output_dir: Target directory (default: cwd).
            extension: Filter by file extension (e.g. ``'pdf'``).

        Returns:
            List of paths to downloaded files.
        """
        import logging

        if output_dir is None:
            output_dir = Path.cwd()
        elif not output_dir.exists():
            output_dir.mkdir(parents=True, exist_ok=True)

        att_list = self.attachments(record_id)

        if extension:
            ext = extension.lower().lstrip(".")
            att_list = [a for a in att_list if a.get("name", "").lower().endswith(f".{ext}")]

        downloaded: builtins.list[Path] = []
        for att_meta in att_list:
            filename = att_meta.get("name", f"attachment_{att_meta['id']}")
            try:
                att_data = self._client.read(
                    "ir.attachment",
                    [att_meta["id"]],
                    self._ATTACHMENT_READ_FIELDS,
                )
                if not att_data:
                    continue
                att = att_data[0]
                filename = att.get("name", f"attachment_{att_meta['id']}")
                out = output_dir / filename
                if att.get("datas"):
                    out.write_bytes(base64.b64decode(att["datas"]))
                    downloaded.append(out)
            except Exception as e:
                logging.getLogger("vodoo").warning("Failed to download %s: %s", filename, e)
                continue
        return downloaded

    def attachment_data(
        self,
        attachment_id: int,
    ) -> bytes:
        """Read an attachment and return raw binary content.

        Args:
            attachment_id: Attachment ID.

        Returns:
            Raw bytes.

        Raises:
            RecordNotFoundError: If attachment not found or has no data.
        """
        from vodoo.base import _decode_attachment_data

        attachments = self._client.read(
            "ir.attachment", [attachment_id], self._ATTACHMENT_READ_FIELDS
        )
        if not attachments:
            raise RecordNotFoundError("ir.attachment", attachment_id)
        return _decode_attachment_data(attachments[0], attachment_id)

    def all_attachment_data(
        self,
        record_id: int,
    ) -> builtins.list[tuple[int, str, bytes]]:
        """Read all attachments for a record, returning in-memory data.

        Args:
            record_id: Record ID.

        Returns:
            List of ``(attachment_id, filename, raw_bytes)`` tuples.
        """
        from vodoo.base import _decode_attachment_record

        att_list = self.attachments(record_id)
        result: builtins.list[tuple[int, str, bytes]] = []
        for att_meta in att_list:
            att_id = att_meta["id"]
            try:
                att_data = self._client.read(
                    "ir.attachment",
                    [att_id],
                    ["id", *self._ATTACHMENT_READ_FIELDS],
                )
                if not att_data:
                    continue
                decoded = _decode_attachment_record(att_data[0], att_id)
                if decoded is not None:
                    result.append(decoded)
            except Exception:
                continue
        return result

    # -- URL -----------------------------------------------------------------

    def url(self, record_id: int) -> str:
        """Return the Odoo web URL for a record.

        Args:
            record_id: Record ID.

        Returns:
            Full URL to the record form view.
        """
        base_url = self._client.config.url.rstrip("/")
        return f"{base_url}/web#id={record_id}&model={self._model}&view_type=form"


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _convert_to_html(text: str, use_markdown: bool = False) -> str:
    """Convert text to HTML, optionally processing markdown."""
    if use_markdown:
        from vodoo.content import _markdown_to_html

        return _markdown_to_html(text)
    return f"<p>{text}</p>"

File: packages/typescript/src/namespaces/domain.ts
import type { OdooClientApi, SearchReadOptions } from "../client-api.js";
import { binaryBytes, decodeBase64, encodeBase64 } from "../binary.js";
import type { BinaryInput } from "../binary.js";
import { richTextToHtml } from "../content.js";
import type { RichText } from "../content.js";
import { decodeRecordDates } from "../dates.js";
import {
  ConfigurationError,
  RecordNotFoundError,
  VodooError,
} from "../errors.js";
import type { OdooRecord } from "../types.js";

const TAG_FIELDS = ["id", "name", "color"] as const;
const MESSAGE_FIELDS = [
  "id",
  "date",
  "author_id",
  "body",
  "subject",
  "message_type",
  "subtype_id",
  "email_from",
] as const;
const ATTACHMENT_LIST_FIELDS = [
  "id",
  "name",
  "file_size",
  "mimetype",
  "create_date",
] as const;
const ATTACHMENT_READ_FIELDS = ["name", "datas", "file_size"] as const;

export interface NamespaceMetadata {
  readonly model: string;
  readonly defaultFields: readonly string[];
  readonly defaultDetailFields?: readonly string[];
  readonly tagModel?: string;
}

export interface ListOptions extends Omit<SearchReadOptions, "fields"> {
  fields?: readonly string[];
}

export interface MessageOptions {
  userId?: number;
  markdown?: boolean;
}

export interface AttachmentOptions {
  mimetype?: string;
}

export interface AttachmentData {
  readonly id: number;
  readonly name: string;
  readonly data: Uint8Array;
  readonly mimetype?: string;
}

function many2OneId(value: unknown): number | null {
  if (typeof value === "number" && Number.isInteger(value)) return value;
  if (Array.isArray(value) && typeof value[0] === "number") return value[0];
  return null;
}

/** Shared namespace CRUD, messaging, tags, and in-memory attachments. */
export class DomainNamespace {
  protected readonly client: OdooClientApi;
  protected readonly metadata: NamespaceMetadata;

  constructor(client: OdooClientApi, metadata: NamespaceMetadata) {
    this.client = client;
    this.metadata = metadata;
  }

  protected decodeRecord(record: OdooRecord): OdooRecord {
    return record;
  }

  async list(options: ListOptions = {}): Promise<OdooRecord[]> {
    const request: SearchReadOptions = {
      fields: options.fields ?? this.metadata.defaultFields,
      limit: options.limit === undefined ? 50 : options.limit,
      order: options.order ?? "create_date desc",
    };
    if (options.domain !== undefined) request.domain = options.domain;
    if (options.offset !== undefined) request.offset = options.offset;
    const records = await this.client.searchRead(this.metadata.model, request);
    return records.map((record) => this.decodeRecord(record));
  }

  async get(recordId: number, fields?: readonly string[]): Promise<OdooRecord> {
    const selectedFields = fields ?? this.metadata.defaultDetailFields;
    const records = await this.client.read(
      this.metadata.model,
      [recordId],
      selectedFields,
    );
    const record = records[0];
    if (record === undefined) {
      throw new RecordNotFoundError(this.metadata.model, recordId);
    }
    return this.decodeRecord(record);
  }

  set(
    recordId: number,
    values: Readonly<Record<string, unknown>>,
  ): Promise<boolean> {
    return this.client.write(this.metadata.model, [recordId], values);
  }

  fields(): Promise<Record<string, unknown>> {
    return this.client.fieldsGet(this.metadata.model);
  }

  comment(
    recordId: number,
    message: RichText,
    options: MessageOptions = {},
  ): Promise<boolean> {
    return this.commentWithId(recordId, message, options).then(
      (messageId) => messageId > 0,
    );
  }

  commentWithId(
    recordId: number,
    message: RichText,
    options: MessageOptions = {},
  ): Promise<number> {
    return this.postMessage(recordId, message, false, options);
  }

  note(
    recordId: number,
    message: RichText,
    options: MessageOptions = {},
  ): Promise<boolean> {
    return this.noteWithId(recordId, message, options).then(
      (messageId) => messageId > 0,
    );
  }

  noteWithId(
    recordId: number,
    message: RichText,
    options: MessageOptions = {},
  ): Promise<number> {
    return this.postMessage(recordId, message, true, options);
  }

  async messages(
    recordId: number,
    limit: number | null = null,
  ): Promise<OdooRecord[]> {
    const records = await this.client.searchRead("mail.message", {
      domain: [
        ["model", "=", this.metadata.model],
        ["res_id", "=", recordId],
      ],
      fields: MESSAGE_FIELDS,
      order: "date desc",
      limit,
    });
    return records.map((record) =>
      decodeRecordDates(record, { date: "datetime" }),
    );
  }

  async tags(): Promise<OdooRecord[]> {
    if (this.metadata.tagModel === undefined) {
      throw new VodooError(`No tag model defined for ${this.metadata.model}`);
    }
    return this.client.searchRead(this.metadata.tagModel, {
      fields: TAG_FIELDS,
      order: "name",
    });
  }

  async addTag(recordId: number, tagId: number): Promise<boolean> {
    const record = await this.get(recordId, ["tag_ids"]);
    const existing = Array.isArray(record.tag_ids)
      ? record.tag_ids.filter(
          (value): value is number => typeof value === "number",
        )
      : [];
    if (existing.includes(tagId)) return true;
    return this.client.write(this.metadata.model, [recordId], {
      tag_ids: [[6, 0, [...existing, tagId]]],
    });
  }

  async attachments(recordId: number): Promise<OdooRecord[]> {
    const records = await this.client.searchRead("ir.attachment", {
      domain: [
        ["res_model", "=", this.metadata.model],
        ["res_id", "=", recordId],
      ],
      fields: ATTACHMENT_LIST_FIELDS,
    });
    return records.map((record) =>
      decodeRecordDates(record, { create_date: "datetime" }),
    );
  }

  /** Attach Worker-safe binary data. Filesystem paths are intentionally unsupported. */
  async attach(
    recordId: number,
    data: BinaryInput,
    name: string,
    options: AttachmentOptions = {},
  ): Promise<number> {
    const bytes = await binaryBytes(data);
    const values: Record<string, unknown> = {
      name,
      datas: encodeBase64(bytes),
      res_model: this.metadata.model,
      res_id: recordId,
      type: "binary",
    };
    const mimetype =
      options.mimetype ??
      (typeof Blob !== "undefined" && data instanceof Blob && data.type !== ""
        ? data.type
        : undefined);
    if (mimetype !== undefined) values.mimetype = mimetype;
    return this.client.create("ir.attachment", values);
  }

  async attachmentData(attachmentId: number): Promise<Uint8Array> {
    const attachments = await this.client.read(
      "ir.attachment",
      [attachmentId],
      ATTACHMENT_READ_FIELDS,
    );
    const attachment = attachments[0];
    if (attachment === undefined) {
      throw new RecordNotFoundError("ir.attachment", attachmentId);
    }
    if (typeof attachment.datas === "string") {
      return decodeBase64(attachment.datas);
    }
    if (attachment.file_size === 0) return new Uint8Array();
    throw new RecordNotFoundError("ir.attachment", attachmentId);
  }

  async allAttachmentData(recordId: number): Promise<AttachmentData[]> {
    const metadata = await this.attachments(recordId);
    const result: AttachmentData[] = [];
    for (const attachment of metadata) {
      if (typeof attachment.id !== "number") continue;
      try {
        const records = await this.client.read(
          "ir.attachment",
          [attachment.id],
          ["id", ...ATTACHMENT_READ_FIELDS],
        );
        const record = records[0];
        if (record === undefined) continue;
        let data: Uint8Array;
        if (typeof record.datas === "string") {
          data = decodeBase64(record.datas);
        } else if (record.file_size === 0) {
          data = new Uint8Array();
        } else {
          continue;
        }
        const item: AttachmentData = {
          id: attachment.id,
          name:
            typeof record.name === "string"
              ? record.name
              : `attachment_${attachment.id}`,
          data,
        };
        if (typeof attachment.mimetype === "string") {
          result.push({ ...item, mimetype: attachment.mimetype });
        } else {
          result.push(item);
        }
      } catch {
        // Match Python all_attachment_data: one unreadable attachment is skipped.
      }
    }
    return result;
  }

  /** Worker-safe replacement for filesystem download: return matching bytes in memory. */
  async downloadAttachments(
    recordId: number,
    extension?: string,
  ): Promise<AttachmentData[]> {
    const attachments = await this.allAttachmentData(recordId);
    if (extension === undefined) return attachments;
    const suffix = `.${extension.replace(/^\./u, "").toLocaleLowerCase("und")}`;
    return attachments.filter((attachment) =>
      attachment.name.toLocaleLowerCase("und").endsWith(suffix),
    );
  }

  /** Python-compatible name for the in-memory Worker download adaptation. */
  download(recordId: number, extension?: string): Promise<AttachmentData[]> {
    return this.downloadAttachments(recordId, extension);
  }

  url(recordId: number): string {
    return `${this.client.url.replace(/\/+$/u, "")}/web#id=${recordId}&model=${this.metadata.model}&view_type=form`;
  }

  private async postMessage(
    recordId: number,
    message: RichText,
    isNote: boolean,
    options: MessageOptions,
  ): Promise<number> {
    const userId = options.userId ?? this.client.defaultUserId;
    if (userId === undefined) {
      throw new ConfigurationError("No default user ID configured");
    }
    const users = await this.client.read("res.users", [userId], ["partner_id"]);
    if (users[0] === undefined) {
      throw new RecordNotFoundError("res.users", userId);
    }
    const partnerId = many2OneId(users[0].partner_id);
    if (partnerId === null) {
      throw new RecordNotFoundError("res.partner", 0);
    }
    const subtypeIds = await this.client.search("mail.message.subtype", {
      domain: [["name", "=", isNote ? "Note" : "Discussions"]],
      limit: 1,
    });
    return this.client.create("mail.message", {
      model: this.metadata.model,
      res_id: recordId,
      body: richTextToHtml(message, options.markdown ?? true),
      message_type: isNote ? "notification" : "comment",
      subtype_id: subtypeIds[0] ?? false,
      author_id: partnerId,
    });
  }
}
